internal/js_library/js_library.bzl

```diff
@@ -49,96 +53,124 @@ def write_amd_names_shim(actions, amd_names_shim, targets):
     actions.write(amd_names_shim, amd_names_shim_content)
 
-def _impl(ctx):
-    files = []
-    typings = []
-    js_files = []
+def _js_library_impl(ctx):
+    direct_sources = depset(ctx.files.srcs)
+    direct_named_module_sources = depset(ctx.files.named_module_srcs)
+    sources_depsets = [direct_sources]
+    named_module_sources_depsets = [direct_named_module_sources]
 
+    include_npm_package_info = False
     for src in ctx.files.srcs:
-        if src.is_source and not src.path.startswith("external/"):
-            dst = ctx.actions.declare_file(src.basename, sibling = src)
-            if ctx.attr.is_windows:
-                copy_cmd(ctx, src, dst)
-            else:
-                copy_bash(ctx, src, dst)
-            if dst.basename.endswith(".d.ts"):
-                typings.append(dst)
-            else:
-                files.append(dst)
-        elif src.basename.endswith(".d.ts"):
-            typings.append(src)
-        else:
-            files.append(src)
-
-    for p in files:
-        if p.basename.endswith(".js") or p.basename.endswith(".js.map") or p.basename.endswith(".json"):
-            js_files.append(p)
-
-    files_depset = depset(files)
+        if src.is_source and src.path.startswith("external/"):
+            include_npm_package_info = True
+            break
+
+    declarations = depset([
+        f
+        for f in ctx.files.srcs
+        if (
+               f.path.endswith(".d.ts") or
+               # package.json may be required to resolve "typings" key
+               f.path.endswith("/package.json")
+           ) and
+           # exclude eg. external/npm/node_modules/protobufjs/node_modules/@types/node/index.d.ts
+           # these would be duplicates of the typings provided directly in another dependency.
+           # also exclude all /node_modules/typescript/lib/lib.*.d.ts files as these are determined by
+           # the tsconfig "lib" attribute
+           len(f.path.split("/node_modules/")) < 3 and f.path.find("/node_modules/typescript/lib/lib.") == -1
+    ])
+
+    transitive_declarations_depsets = [declarations]
+
+    for dep in ctx.attr.deps:
+        if DeclarationInfo in dep:
+            transitive_declarations_depsets.append(dep[DeclarationInfo].transitive_declarations)
+        if NpmPackageInfo in dep:
+            sources_depsets.append(dep[NpmPackageInfo].sources)
+        if JSModuleInfo in dep:
+            sources_depsets.append(dep[JSModuleInfo].sources)
+        if JSNamedModuleInfo in dep:
+            named_module_sources_depsets.append(dep[JSNamedModuleInfo].sources)
+
+    transitive_declarations = depset(transitive = transitive_declarations_depsets)
+    transitive_sources = depset(transitive = sources_depsets + named_module_sources_depsets)
 
     providers = [
         DefaultInfo(
-            files = files_depset,
-            runfiles = ctx.runfiles(files = ctx.files.srcs),
+            files = direct_sources,
+        ),
+        DeclarationInfo(
```

Comment:
does this produce DeclarationInfo when there are no .d.ts files? if so that's not nice as it takes away the error messaging when you put it in the deps of a ts_project

Reply:
nice one 👍 fixed